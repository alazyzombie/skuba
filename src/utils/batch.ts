const createBatches = <T extends unknown>(
  list: T[],
  batchSize: number,
): T[][] => {
  if (!Number.isInteger(batchSize) || batchSize < 0) {
    throw new Error('Invalid batchSize provided');
  }

  const batchList: T[][] = [];
  for (let i = 0; i < Math.ceil(list.length / batchSize); i++) {
    const startIndex = i * batchSize;
    batchList.push(list.slice(startIndex, startIndex + batchSize));
  }
  return batchList;
};

export { createBatches };
