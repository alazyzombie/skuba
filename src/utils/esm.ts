import path from 'path';
import { fileURLToPath } from 'url';

export const dirname = (importMeta: ImportMeta) => {
  const __filename = fileURLToPath(importMeta.url);

  return path.dirname(__filename);
};
