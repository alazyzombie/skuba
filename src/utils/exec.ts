import util from 'util';

import chalk from 'chalk';
import concurrently, { CommandObj } from 'concurrently';
import execa from 'execa';
import npmRunPath from 'npm-run-path';
import npmWhich from 'npm-which';

import { isErrorWithCode } from './error';

interface ExecConcurrentlyCommand {
  command: string;
  name: string;

  /**
   * Whether to try to resolve the npm binary through a shim if skuba detects
   * that it is running in a Yarn Plug'n'Play environment.
   *
   * npm binaries can't be located through conventional `node_modules/.bin`
   * means with PnP.
   */
  pnp?: boolean;
}

type ExecOptions = execa.Options & {
  /**
   * Whether to try to resolve the command through a shim if skuba detects that
   * it is running in a Yarn Plug'n'Play environment.
   *
   * Enable this for npm "binaries"; they can't be located through conventional
   * `node_modules/.bin` means when PnP is enabled.
   */
  pnp?: boolean;

  streamStdio?: true;
};

const NPM_RUN_PATH = npmRunPath({ cwd: __dirname });

const withPnpShim = (
  isEnabled: boolean | undefined,
  command: string,
  args: string[],
) => {
  if (!isEnabled || !process.versions.hasOwnProperty('pnp')) {
    return { command, args };
  }

  const shimPath = require.resolve('./pnp');

  return {
    command: 'node',
    args: [shimPath, command, ...args],
  };
};

const runCommand = (command: string, args: string[], opts?: ExecOptions) => {
  const resolved = withPnpShim(opts?.pnp, command, args);

  const subprocess = execa(resolved.command, resolved.args, {
    localDir: __dirname,
    preferLocal: true,
    stdio: 'inherit',
    ...opts,
  });

  if (opts?.streamStdio === true) {
    subprocess.stderr?.pipe(process.stderr);
    subprocess.stdout?.pipe(process.stdout);
  }

  return subprocess;
};

const whichCallback = npmWhich(__dirname);

const which = util.promisify<string, string>(whichCallback);

export const createExec = (opts: ExecOptions) => (
  command: string,
  ...args: string[]
) => runCommand(command, args, opts);

export const exec = async (command: string, ...args: string[]) =>
  runCommand(command, args);

export const execConcurrently = (commands: ExecConcurrentlyCommand[]) =>
  concurrently(
    commands.map(({ command, pnp, ...rest }) => {
      const resolved = withPnpShim(pnp, command, []);

      const obj: CommandObj = {
        ...rest,
        command: [resolved.command, ...resolved.args].join(' '),
        env: pnp ? undefined : { PATH: NPM_RUN_PATH },
      };

      return obj;
    }),
  );

export const ensureCommands = async (...names: string[]) => {
  let success = true;

  await Promise.all(
    names.map(async (name) => {
      try {
        return await which(name);
      } catch (err) {
        if (isErrorWithCode(err, 'ENOENT')) {
          success = false;

          return console.error(chalk.bold(name), 'needs to be installed.');
        }

        throw err;
      }
    }),
  );

  if (!success) {
    process.exit(1);
  }
};
