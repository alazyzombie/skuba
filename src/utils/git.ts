import path from 'path';

import fs from 'fs-extra';
import git from 'isomorphic-git';
import http from 'isomorphic-git/http/node';

// Status Matrix Row Indexes
const FILEPATH = 0;
const HEAD = 1;
const WORKDIR = 2;
const STAGE = 3;

// Status Matrix State
const ABSENT = 0;
const UNCHANGED = 1;

type GitUser = NonNullable<Parameters<typeof git.commit>[0]['author']>;

export const gitCommit = async ({
  author = {
    name: 'skuba',
  },
  committer = {
    name: 'skuba',
  },
  dir,
  message,
}: {
  author?: GitUser;
  committer?: GitUser;
  dir: string;
  message: string;
}) =>
  git.commit({
    author,
    committer,
    dir,
    fs,
    message,
  });

interface GitPushProps {
  auth: { type: 'gitHubApp'; token: string };
  branch?: string;
  commitOid: string;
  dir: string;
  force?: boolean;
}

export const gitPush = async ({
  auth,
  branch,
  commitOid,
  dir,
  force,
}: GitPushProps) => {
  const { owner, repo } = await getOwnerRepo(dir);

  const url = `https://github.com/${encodeURIComponent(
    owner,
  )}/${encodeURIComponent(repo)}`;

  return git.push({
    onAuth: () => ({
      username: 'x-access-token',
      password: auth.token,
    }),
    http,
    dir,
    fs,
    ref: commitOid,
    remoteRef: branch,
    url,
    force,
  });
};

export const getHeadSha = async (dir: string): Promise<string> => {
  const [commit] = await git.log({ depth: 1, dir, fs });

  return commit.oid;
};

/**
 * Matches the owner and repository names in a GitHub repository URL.
 *
 * For example, given the following input strings:
 *
 * ```console
 * git@github.com:seek-oss/skuba.git
 * https://github.com/seek-oss/skuba.git
 * ```
 *
 * This pattern will produce the following matches:
 *
 * 1. seek-oss
 * 2. skuba
 */
const ownerRepoRegex = /github.com(?::|\/)(.+)\/(.+).git$/;

export const getOwnerRepo = async (
  dir: string,
): Promise<{ owner: string; repo: string }> => {
  const remotes = await git.listRemotes({ dir, fs });

  for (const { url } of remotes) {
    const match = ownerRepoRegex.exec(url);

    const owner = match?.[1];
    const repo = match?.[2];

    if (owner && repo) {
      return { owner, repo };
    }
  }

  throw new Error('Could not find a GitHub remote');
};

export const getCurrentBranchRef = async (dir: string): Promise<string> => {
  const branch = await git.currentBranch({
    fs,
    dir,
    fullname: true,
  });

  if (!branch) {
    throw new Error('Could not find the current branch');
  }

  return branch;
};

export const gitBranch = async ({
  dir,
  ref,
  checkout,
}: {
  dir: string;
  ref: string;
  checkout?: boolean;
}): Promise<void> => {
  await git.branch({
    fs,
    dir,
    ref,
    checkout,
  });
};

export const gitDeleteBranch = async ({
  dir,
  ref,
}: {
  dir: string;
  ref: string;
}): Promise<void> => {
  await git.deleteBranch({
    fs,
    dir,
    ref,
  });
};

export const gitListTags = async ({
  dir,
}: {
  dir: string;
}): Promise<string[]> =>
  git.listTags({
    fs,
    dir,
  });

export interface ChangedFile {
  path: string;
  deleted: boolean;
}

/**
 * Returns files which have been changed
 */
export const getChangedFiles = async ({
  dir,
}: {
  dir: string;
}): Promise<ChangedFile[]> => {
  const allFiles = await git.statusMatrix({ fs, dir });
  return allFiles
    .filter(
      (row) =>
        row[HEAD] !== UNCHANGED ||
        row[WORKDIR] !== UNCHANGED ||
        row[STAGE] !== UNCHANGED,
    )
    .map((row) => ({ path: row[FILEPATH], deleted: row[WORKDIR] === ABSENT }));
};

export const gitAdd = async ({
  dir,
  filepath,
}: {
  dir: string;
  filepath: string;
}): Promise<void> => {
  await git.add({ fs, dir, filepath });
};

export const gitRemove = async ({
  dir,
  filepath,
}: {
  dir: string;
  filepath: string;
}): Promise<void> => {
  await git.remove({ fs, dir, filepath });
};

export const gitReset = async ({
  dir,
  branch,
  commitOid,
  hard,
}: {
  dir: string;
  branch: string;
  commitOid: string;
  hard?: boolean;
}): Promise<void> => {
  await fs.promises.writeFile(
    path.join(dir, '.git/refs/heads', branch),
    `${commitOid}\n`,
  );

  if (hard) {
    const allFiles = await git.statusMatrix({ dir, fs });
    // Get all files which have been modified or staged - does not include new untracked files or deleted files
    const modifiedFiles = allFiles
      .filter((row) => row[WORKDIR] > UNCHANGED && row[STAGE] > UNCHANGED)
      .map((row) => row[FILEPATH]);

    // Delete modified/staged files
    await Promise.all(
      modifiedFiles.map((filePath) => fs.promises.rm(filePath)),
    );
    await git.checkout({ dir, fs, ref: branch, force: true });
  }
};

export const gitCommitAllChanges = async ({
  dir,
  message,
  author = {
    name: 'skuba',
  },
  committer = {
    name: 'skuba',
  },
}: {
  dir: string;
  message: string;
  author?: GitUser;
  committer?: GitUser;
}) => {
  const changedFiles = await getChangedFiles({ dir });
  await Promise.all(
    changedFiles.map((file) =>
      !file.deleted
        ? gitAdd({ dir, filepath: file.path })
        : gitRemove({ dir, filepath: file.path }),
    ),
  );

  await gitCommit({
    dir,
    message,
    author,
    committer,
  });
};
