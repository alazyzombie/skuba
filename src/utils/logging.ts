/* eslint-disable no-console */

import chalk from 'chalk';

export const log = {
  bold: (...message: unknown[]) => chalk.bold(...message),
  formatSubtle: (...message: unknown[]) => chalk.grey(...message),

  subtle: (...message: unknown[]) => console.log(chalk.grey(...message)),
  err: (...message: unknown[]) => console.error(chalk.red(...message)),
  newline: () => console.log(),
  ok: (...message: unknown[]) => console.log(chalk.green(...message)),
  plain: (...message: unknown[]) => console.log(...message),
  warn: (...message: unknown[]) => console.error(chalk.yellow(...message)),
};
