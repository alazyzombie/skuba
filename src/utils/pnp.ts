import chalk from 'chalk';
import { createRequire } from 'module';
import path from 'path';

import { isObject, isObjectWithStringProp } from './validation';

const BINARY_REMAPS: Record<string, string> = {
  typescript: 'tsc',
};

const localRequire = createRequire(__filename);

const createRequireRelative = (packageName: string) => (pathname: string) => {
  const resolved = localRequire.resolve(path.join(packageName, pathname));

  return localRequire(resolved) as unknown;
};

const fail = (message: string) => {
  console.error(chalk.red(message));
  process.exit(1);
};

const getPartialBinPath = (
  packageName: string,
  packageJson: unknown,
): string => {
  if (!isObject(packageJson)) {
    return fail(`${chalk.bold(`${packageName}/package.json`)} not found`);
  }

  if (typeof packageJson.bin === 'string') {
    return packageJson.bin;
  }

  const binName = BINARY_REMAPS[packageName] ?? packageName;

  if (isObjectWithStringProp(packageJson.bin, binName)) {
    return packageJson.bin[binName];
  }

  return fail(
    `${chalk.bold(`${packageName}/package.json#/bin`)} not recognised`,
  );
};

const runBinary = (packageName: string) => {
  const requireRelative = createRequireRelative(packageName);

  // Derive the binary path from package.json#/bin.
  const packageJson = requireRelative('package.json');
  const partialBinPath = getPartialBinPath(packageName, packageJson);

  // Require the binary.
  return requireRelative(partialBinPath);
};

// Hide shim argument from binary.
const [packageName] = process.argv.splice(2);

runBinary(packageName);
