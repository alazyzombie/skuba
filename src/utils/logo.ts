import chalk from 'chalk';
import isInstalledGlobally from 'is-installed-globally';

import { log } from './logging.js';
import { getSkubaVersionInfo } from './version.js';

const LOGO = chalk.blueBright(`
    ╭─╮ ${chalk.magentaBright('    ')}╭─╮
╭───│ ╰─${chalk.magentaBright('╭─┬─╮')} ╰─╮───╮
│_ ─┤  <${chalk.magentaBright('│ ╵ │')} • │ • │
╰───╰─┴─${chalk.magentaBright('╰───╯')}───╯── ╰
`);

export const showLogoAndVersionInfo = async () => {
  const versionInfo = await getSkubaVersionInfo();

  log.plain(LOGO);
  log.subtle(
    log.bold(versionInfo.local),
    '|',
    'latest',
    log.bold(versionInfo.latest ?? 'offline ✈'),
  );
  log.newline();

  if (versionInfo.isStale) {
    log.warn('Your skuba installation is out of date.');
    log.warn('Consider upgrading:');
    log.newline();
    log.warn(
      log.bold(
        'yarn',
        ...(isInstalledGlobally ? ['global'] : []),
        'upgrade',
        `skuba@${versionInfo.latest}`,
      ),
    );
    log.newline();
  }

  return versionInfo;
};
