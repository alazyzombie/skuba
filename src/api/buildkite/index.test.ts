import * as execModule from '../../utils/exec';

import { annotate } from '.';

const exec = jest.spyOn(execModule, 'exec');

beforeEach(() => {
  jest.clearAllMocks();

  exec.mockResolvedValue(undefined as any);
});

afterEach(() => {
  delete process.env.BUILDKITE;
  delete process.env.BUILDKITE_AGENT_ACCESS_TOKEN;
  delete process.env.BUILDKITE_JOB_ID;
  delete process.env.BUILDKITE_STEP_ID;
});

const setEnvironmentVariables = () => {
  process.env.BUILDKITE = 'true';
  process.env.BUILDKITE_AGENT_ACCESS_TOKEN = 'token';
  process.env.BUILDKITE_JOB_ID = 'job-id';
  process.env.BUILDKITE_STEP_ID = 'step-id';
};

describe('annotate', () => {
  const markdown = '**Message**';

  describe.each`
    description                    | opts
    ${'undefined options'}         | ${undefined}
    ${'empty options'}             | ${{}}
    ${'context option'}            | ${{ context: 'skuba-lint' }}
    ${'style option'}              | ${{ style: 'info' }}
    ${'context and style options'} | ${{ context: 'skuba-test', style: 'error' }}
  `('with $description', ({ opts }) => {
    it('annotates when environment variables are set', async () => {
      setEnvironmentVariables();

      await expect(annotate(markdown, opts)).resolves.toBeUndefined();

      expect(
        [null, ...exec.mock.calls.flat(), null].join('\n'),
      ).toMatchSnapshot();
    });

    it('skips when environment variables are not set', async () => {
      await expect(annotate(markdown)).resolves.toBeUndefined();

      expect(exec).not.toHaveBeenCalled();
    });
  });
});
