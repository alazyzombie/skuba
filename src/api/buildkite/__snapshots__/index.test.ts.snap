// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotate with context and style options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
step-id-skuba-test
--style
error
**Message**
"
`;

exports[`annotate with context option annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
step-id-skuba-lint
**Message**
"
`;

exports[`annotate with empty options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
step-id
**Message**
"
`;

exports[`annotate with style option annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
step-id
--style
info
**Message**
"
`;

exports[`annotate with undefined options annotates when environment variables are set 1`] = `
"
buildkite-agent
annotate
--context
step-id
**Message**
"
`;
